aws-cpp-sdk-core-tests/utils/threading/ExecutorTest.cpp

```diff
@@ -0,0 +1,275 @@
+//                OnIsolatedTaskComplete();
+//            };
+//            
+//            std::thread t(isolatedFn);
+//            t.detach();
+//        }
+//        else
+//        {
+//            std::thread t(fn);
+//            t.detach();
+//        }
+//        return true;
+//    }
+//    
+//    void OnIsolatedTaskComplete()
+//    {
+//        std::unique_lock<std::mutex> locker(m_isolatedTaskLock);
+//        m_isolatedTaskRunning = false;
+//        locker.unlock();
+//        m_isolatedTaskDoneSignal.notify_all();
+//    }
+//    
+//private:
+//    bool m_createIsolatedTask = false;
+//    bool m_isolatedTaskRunning = false;
+//    std::mutex m_isolatedTaskLock;
+//    std::condition_variable m_isolatedTaskDoneSignal;
+//    
+//};
+
+/** Test whether a task gets blocked properly if max pool size is exceeded */
+TEST(BlockingExecutorTest, TestBlockSingleTask)
+{
+    AWS_BEGIN_MEMORY_TEST(16, 10)
+    
+//    std::shared_ptr<MockExecutor> mockExecutor = Aws::MakeShared<MockExecutor>(ALLOCATION_TAG);
+//    BlockingExecutor blockingExecutor(mockExecutor, 3);
+    
+    std::shared_ptr<DefaultExecutor> mockExecutor = Aws::MakeShared<DefaultExecutor>(ALLOCATION_TAG);
+    BlockingExecutor blockingExecutor(mockExecutor, 3);
+    
+    std::vector<u_long> taskOrder;
+    size_t FINAL_SIZE = 4;
+    
+    std::mutex taskLock;
+    std::condition_variable startSignal;
+    std::condition_variable taskSignal;
+    std::condition_variable doneSignal;
+    
+    const u_int8_t WAIT_FOR_START_SIGNAL = 1 << 0;
+    const u_int8_t WAIT_FOR_TASK_SIGNAL = 1 << 1;
+    const u_int8_t SIGNAL_OTHER_TASKS = 1 << 2;
+    
+    auto executeTask = [&](u_long id, u_int8_t flags)
+    {
+        std::unique_lock<std::mutex> locker(taskLock);
+        
+        if ((flags & WAIT_FOR_START_SIGNAL) == WAIT_FOR_START_SIGNAL)
+        {
+            startSignal.wait(locker);
+        }
+        if ((flags & WAIT_FOR_TASK_SIGNAL) == WAIT_FOR_TASK_SIGNAL)
+        {
+            taskSignal.wait(locker);
+        }
+//        // TODO: This tight coupling with the task needing to access the executor
+//        // is probably not the best code architecture.
+//        std::unique_lock<std::mutex> isolatedLocker(mockExecutor->)
+//        if (mockExecutor->IsIsolatedTaskRunning())
+//        {
+//            
+//        }
+        taskOrder.emplace_back(id);
+        if (taskOrder.size() == FINAL_SIZE)
+        {
+            locker.unlock();
+            doneSignal.notify_all();
+        }
+        else
+        {
+            if ((flags & SIGNAL_OTHER_TASKS) == SIGNAL_OTHER_TASKS)
+            {
+                locker.unlock();
+                taskSignal.notify_all();
+                locker.lock();
+            }
+        }
+    };
+    
+    // Tasks 0 and 1 wait until they are signaled by another task
+    blockingExecutor.Submit(executeTask, 0, WAIT_FOR_TASK_SIGNAL);
+    blockingExecutor.Submit(executeTask, 1, WAIT_FOR_TASK_SIGNAL);
```

Comment:
Ok, you've got the right idea here but it seems way too complicated:

Having both of the first two tasks wait on a semaphore is good.
The next problem is that if the third task is waiting on the blocked two threads, then it will never execute and according to the contract, it should block the calling thread.

Why not try instead of using the DefaultExecutor for the nested executor, actually mock out the executor as planned (you can certainly borrow the code from DefaultExecutor) moving the semaphore logic in there. Then you can actually control the flow of the test. Embeddeding the semaphore signalling in the tasks itself won't be of much help.

Blocking Executor
    Underlying Executor
           --> semaphore logic in here.

I think you just need one semaphore altogether.
When you call submit the second time, it should be stopped on the signal. so far so good.

now when you call it a third time, you need to call it in a separate thread. std::thread.
You should verify after spawning the thread that execution hasn't happened (a flag check would do the trick).
Then signal the semaphore, and make sure the thread does then execute.